keyedcache/api.go

```diff
@@ -0,0 +1,153 @@
+//
+//   o Equal() -- returns true if the two keys refer to the same object
+//   o Hash() -- return a hash value based on the key (two keys that are equal
+//     must have the same hash value)
+//   o SizeOf() -- return memory used by the key (to track memory consumption);
+//     all keys are assumed to be the same size.
+//
+type Key interface {
+	Equal(key *Key) bool
+	Hash() uint64
+	SizeOf() uint64
+}
+
+// Objects in the cache are "allocated" by calling a NewObject function supplied
+// by the client when it created the cache.  New objects do not have an
+// indentity until they are assigned one by a lookup in the cache.
+//
+// Objects must support the following operations.  These methods are used
+// exclusively by the keyedcache; clients who get an object from the cache
+// should not call them, instead they should call methods provided by the cache
+// to flush or invalidate objects:
+//
+//   o Read() -- called to initialize an object when its assigned an identity;
+//     it can block arbitrarity but cannot call into this instance of the
+//     cache.
+//   o Write() -- called to flush the contents of a dirty object; it can
+//     block arbitrarity but cannot call into this instance of the cache.
+//   o Inval() -- called to destroy the identity of an object before
+//     assigning it a new identity; it can block indefinitely but cannot
+//     call back into this instance of the cache (key is the old identity
+//     of the object)
+//   o Free() -- called when the cache is no longer referencing the object;
+//     it cannot call back into this instance of the cache (I'm not sure that
+//     this method makes sense in Go).
+//   o SizeOf() -- return memory used by the object (to track memory consumption);
+//     all objects are assumed to be the same size.
+//
+// Note: perhaps the Read() and Write() operations should be asynchronous.
+//
+type Object interface {
+	Read(cntxt *ReqContext, key *Key) (err error)
+	Write(cntxt *ReqContext, key *Key) (err error)
+	Inval(cntxt *ReqContext, key *Key)
+	Free(cntxt *ReqContext)
+	SizeOf() uint64
+}
+
+// CachedObject holds the information that a keyedcache uses to manage an object
+// stored in the cache.  An object managed by the cache must include an embedded
+// pointer to this interface, which it must initializ with a pointer passed to
+// the NewObject function supplied by the client of the cache (so when an object
+// is created by NewObject() this interface is initialized).
+//
+// CachedObject supports the following methods. These may be invoked by a client
+// that has a hold on the object.  Successful lookup of an object in the cache
+// returns a pointer to a held object.
+//
+//   o Hold() -- get an additional hold on the object
+//   o Release() -- release a hold on the object; once all holds on an object
+//     are released the object may be invalidated
+//   o MarkDirty() -- mark a held object dirty; dirty objects are cleaned
```

Comment:
I guess we'd require a resource be "held" when calling MarkDirty()...
An alternative might be for the call to Release() to pass back an indication of "dirtiness", but this would mean others wouldn't know it was already dirty until potentially log after it was. Plus, the caller to Release() would be required to "buffer" the knowledge of dirtiness until it was ready to Release() it. I'm fine either way though.